Implement block transfer interface of IPMI

SergiiDmytruk opened the pull request:
As part of `coreboot/src/drivers/ipmi`.

krystian-hebel commented on src/drivers/ipmi/ipmi_kcs.h:
```diff
@@ -3,27 +3,9 @@
-#define   IPMI_APP_SELFTEST_NOT_IMPLEMENTED      0x56
-#define   IPMI_APP_SELFTEST_ERROR                0x57
-#define   IPMI_APP_SELFTEST_FATAL_HW_ERROR       0x58
+#include <stdint.h>
```
There are no changes requiring this include in ipmi_kcs.h, so maybe move it to impmi_if.h

SergiiDmytruk replied:
But there is
```c
void ipmi_bmc_version(uint8_t *ipmi_bmc_major_revision, uint8_t *ipmi_bmc_minor_revision);
```
below from before.

krystian-hebel replied:
Ok, so that is an issue in the current code I guess, let's leave it then.

krystian-hebel commented on src/drivers/ipmi/ipmi_bt.c:
```diff
@@ -0,0 +1,212 @@
+enum cb_err ipmi_bt_premem_init(const uint16_t port, const uint16_t device)
+{
+	return ipmi_premem_init(port, device);
+}
```
Why are those crossed, i.e. first one is generic function name calling BT-specific, and the second the other way around? IMHO we could use just the generic name for both protocols, instead of using wrappers and let the rest for linker - at the moment it is impossible to have both KCS and BT, and I doubt anyone will need both in one FW image. On the other hand, this is something that may have to be revised after review upstream so I won't be forcing any specific approaches, other than keeping the naming scheme consistent.

SergiiDmytruk replied:
To answer your question, `ipmi_message` and `ipmi_bt_premem_init` are both new functions and they redirect to some other code. But actually I just forgot to look how many times `ipmi_kcs_message` is called and update all the places before dropping thunks...

krystian-hebel commented on src/drivers/ipmi/ipmi_bt.c:
```diff
@@ -0,0 +1,212 @@
+	return 0;
+}
+
+int ipmi_bt_ignore(uint16_t port)
```
I think 'clear' or 'reset' would better describe what this function does

krystian-hebel commented on src/drivers/ipmi/Makefile.inc:
```diff
@@ -1,8 +1,18 @@
+ramstage-$(CONFIG_IPMI_BT) += ipmi_ops.c
+ramstage-$(CONFIG_IPMI_BT) += ipmi_fru.c
+romstage-$(CONFIG_IPMI_BT_ROMSTAGE) += ipmi_ops_premem.c
+ramstage-$(CONFIG_IPMI_BT_ROMSTAGE) += ipmi_bt.c
```
```suggestion
romstage-$(CONFIG_IPMI_BT_ROMSTAGE) += ipmi_bt.c
```

krystian-hebel commented on src/soc/ibm/power9/chip.c:
```diff
@@ -183,16 +146,22 @@ static void activate_slave_cores(void)
+	 * Skiboot enables interrupts, but because those are triggered on 0->1
+	 * transition and bit is already set, they do not arrive.
 	 *
-	 * This MUST be done as late as possible so that none of the newly
-	 * activated threads start execution before current thread jumps into
-	 * the payload.
```
Please leave this comment immediately above activate_slave_cores() so nobody tries to add code after it.

SergiiDmytruk replied:
Right, I removed it accidentally during rebase.

krystian-hebel commented on src/soc/ibm/power9/chip.c:
```diff
@@ -183,16 +146,22 @@ static void activate_slave_cores(void)
 void platform_prog_run(struct prog *prog)
 {
-	clear_ipmi_attn();
+	enum { IPMI_BT_PORT_BASE = 0xE4 };
```
I really don't like the idea of having to hardcode it here. I don't know if coreboot gives a way of getting `dev->path.pnp.port` for arbitrary dev, but even if that isn't possible we can have `extern uint16_t ipmi_port` in ipmi_bt.h and set it in `dev->path.pnp.port()`. Another option is to not pass the port address to `ipmi_bt_ignore()` and have a static global variable for port in ipmi_bt.c instead.

SergiiDmytruk replied:
There is a list of all devices, however identification is done by comparing `chip_ops->name`, because I didn't find anything better.

krystian-hebel replied:
Looks good, but please add check for `dev->chip_ops->initialized` and `dev->enabled`, just in case.

macpijan commented:
@SergiiDmytruk All review comments seems to be resolved. One conflict appeared. Please resolve that.

SergiiDmytruk commented:
@macpijan PR wasn't approved, @krystian-hebel probably haven't tested latest changes yet.

krystian-hebel commented on src/drivers/ipmi/ipmi_bt_ops.c:
```diff
@@ -0,0 +1,100 @@
+		dev->enabled = 0;
+	}
+
+	if (!ipmi_bt_clear(dev->path.pnp.port))
```
```suggestion
	if (ipmi_bt_clear(dev->path.pnp.port))
```

krystian-hebel commented on src/drivers/ipmi/ipmi_ops_premem.c:
```diff
@@ -0,0 +1,65 @@
+		return CB_ERR;
+
+#if CONFIG(IPMI_BT)
+	if (!ipmi_bt_clear(dev->path.pnp.port))
```
```suggestion
	if (ipmi_bt_clear(dev->path.pnp.port))
```

krystian-hebel reviewed (changes requested):
Either change as in suggestions or reverse return values for bt_clear and change the condition checked in chip.c

SergiiDmytruk committed:
drivers/ipmi: prepare for adding more interfaces

Deduplicate common initialization code (self-test and device
identification) and put it in a new ipmi_if.c unit, which is
supposed to work with any underlying IPMI interface.

Change-Id: Ia99da6fb63adb7bf556d3d6f7964b34831be8a2f
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
drivers/ipmi: add BT interface

Change-Id: Idb67972d1c38bbae04c7b4de3405350c229a05b9
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
soc/power9/timer.c: add udelay() with active wait

Change-Id: I90f3b00c79420bd9d19c2202ee76daa0d3944ae0
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

SergiiDmytruk committed:
soc/power9/chip.c: use IPMI BT driver

Change-Id: Ia73a132a724be2c2f9ec7398e4bcae0372c0e715
Signed-off-by: Sergii Dmytruk <sergii.dmytruk@3mdeb.com>

krystian-hebel commented:
I am doing rebase and merge because commits are already nicely split for upstreaming.
